buildtools/npm/manifest.go

```diff
@@ -35,3 +65,102 @@ type Lockfile struct {
 func FromLockfile(filename string) (Lockfile, error) {
 	return Lockfile{}, errors.ErrNotImplemented
 }
+
+// TODO: add support for NODE_PATH and GLOBAL_FOLDERS.
+func subModulePath(moduleName string, currentDir string) (string, error) {
+	var modulePath string
+	parentPath := filepath.Dir(currentDir)
+	if filepath.Base(currentDir) == "node_modules" {
+		modulePath = filepath.Join(currentDir, moduleName)
+
+		// remove extra node_modules
+		parentPath = filepath.Dir(parentPath)
+	} else {
+		modulePath = filepath.Join(currentDir, "node_modules", moduleName)
+	}
+
+	moduleFolderExists, err := files.ExistsFolder(modulePath)
+	if err != nil {
+		return "", err
+	}
+
+	if moduleFolderExists {
+		return modulePath, nil
+	}
+
+	if parentPath == currentDir {
+		return "", errors.New("cannot find source of module " + moduleName)
+	}
+
+	return subModulePath(moduleName, parentPath)
+}
+
+func convertManifestToPkg(manifest manifest) pkg.Package {
+	id := pkg.ID{
+		Type:     pkg.NodeJS,
+		Name:     manifest.Name,
+		Revision: manifest.Version,
+	}
+
+	var imports pkg.Imports
+	for depName, version := range manifest.Dependencies {
+		id := pkg.ID{
+			Type:     pkg.NodeJS,
+			Name:     depName,
+			Revision: version,
+		}
+
+		depImport := pkg.Import{
+			Target:   depName,
+			Resolved: id,
+		}
+		imports = append(imports, depImport)
+	}
+
+	return pkg.Package{
+		ID:      id,
+		Imports: imports,
+	}
+}
+
+func fromSubNodeModules(currentDir string, previousPackage pkg.Package) (map[pkg.ID]pkg.Package, error) {
+	submoduleProjects := make(map[pkg.ID]pkg.Package)
+
+	for i, imported := range previousPackage.Imports {
+		pathToSubModule, subProject, err := subModule(imported.Target, currentDir)
+		if err != nil {
+			return nil, err
+		}
+
+		submoduleProjects[subProject.ID] = subProject
+
+		// update previous project's revision resolved reference to stamp out non-deterministic behavior (e.g. semver defined versions in package.json)
+		previousPackage.Imports[i].Resolved.Revision = subProject.ID.Revision
```

Comment:
I'm not certain that this actually _changes_ the imports of the previous package, because the package is passed by value.

Reply:
It does, it's value is referencing an array on the heap, which does not get copied.